server/src/main/java/io/druid/server/coordination/ZkCoordinator.java

```diff
@@ -179,74 +186,78 @@ public void addSegment(DataSegment segment, DataSegmentChangeCallback callback)
-                  @Override
-                  public Void call() throws SegmentLoadingException
-                  {
+        loadingExecutor.submit(
+            new Runnable() {
+              @Override
+              public void run() {
+                try {
+                  log.info("Loading segment[%d/%d][%s]", counter.getAndIncrement(), numSegments, segment.getIdentifier());
+                  final boolean loaded = loadSegment(segment, callback);
+                  if (loaded) {
                     try {
-                      log.info("Loading segment %s", segment.getIdentifier());
-                      final boolean loaded = loadSegment(segment, callback);
-                      if (loaded) {
-                        try {
-                          backgroundSegmentAnnouncer.announceSegment(segment);
-                        }
-                        catch (InterruptedException e) {
-                          Thread.currentThread().interrupt();
-                          throw new SegmentLoadingException(e, "Loading Interrupted");
-                        }
-                      }
-                      return null;
-                    } catch(SegmentLoadingException e) {
-                      log.error(e, "[%s] failed to load", segment.getIdentifier());
-                      throw e;
+                      backgroundSegmentAnnouncer.announceSegment(segment);
+                    } catch (InterruptedException e) {
+                      Thread.currentThread().interrupt();
+                      throw new SegmentLoadingException(e, "Loading Interrupted");
                     }
                   }
+                } catch (SegmentLoadingException e) {
+                  log.error(e, "[%s] failed to load", segment.getIdentifier());
+                  failedSegments.add(segment);
+                } finally {
+                  latch.countDown();
                 }
-            )
+              }
+            }
         );
       }
 
-      int failed = 0;
-      for(ListenableFuture future : segmentLoading) {
-        try {
-          future.get();
-        } catch(InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw new SegmentLoadingException(e, "Loading Interrupted");
-        } catch(ExecutionException e) {
-          failed++;
+      try{
+        latch.await();
+
+        if(failedSegments.size() > 0) {
+          log.makeAlert("%,d errors seen while loading segments", failedSegments.size())
+              .addData("failedSegments", failedSegments);
         }
-      }
-      if(failed > 0) {
-        throw new SegmentLoadingException("%,d errors seen while loading segments", failed);
+      } catch(InterruptedException e) {
+        Thread.currentThread().interrupt();
+        log.makeAlert(e, "LoadingInterrupted");
       }
 
       backgroundSegmentAnnouncer.finishAnnouncing();
     }
     catch (SegmentLoadingException e) {
       log.makeAlert(e, "Failed to load segments")
```

Comment:
I'm fine removing the list of segments, but can we add a count, say `[%d/%d]` segments to log message, similar to what we do for logs above?